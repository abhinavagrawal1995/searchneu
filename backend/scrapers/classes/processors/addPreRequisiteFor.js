/*
 * Copyright (c) 2017 Edward Shen
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import BaseProcessor from './baseProcessor';
import Keys from '../../../../common/Keys';

class AddPreRequisiteFor extends BaseProcessor.BaseProcessor {
  termDump = {}
  classMap = {};

  go(termDump) {
    this.termDump = termDump;
    for (let aClass of this.termDump.classes) {
      console.log(`prereqfor: ${aClass}`);
      console.log(`prereqfor: ${aClass.prereqs}`);
      if (aClass.prereqs) {
        aClass = this.parsePreReqs(aClass, aClass.prereqs);
      }
    }

    for (const aClass of termDump.classes) {
      const key = Keys.create(aClass).getHash();
      this.classMap[key] = aClass;
    }

    return termDump;
  }

  // Recursively traverse the prerequsite structure
  parsePreReqs(mainClass, node, isRequired) {

    if (this.isClass(node)) {
      const find = Keys.create({
        host: mainClass.host,
        termId: mainClass.termId,
        subject: node.subject,
        classUid: node.classUid,
      }).getHash();

      const nodeRef = this.termDump[find];

      if (nodeRef.prereqsFor === undefined) {
        nodeRef.optPrereqsFor = [];
      } else {
        nodeRef.optPrereqsFor.push({
          subject: mainClass.subject,
          classId: mainClass.classId,
        });
      }
    } else {
      const classType = node.type;

      if (node.values !== undefined) {
        node.values.map((course) => {
          // returns if the
          const reqType = (classType === 'and') ? isRequired : false;
          return this.parsePreReqs(mainClass, course, reqType);
        });
      }
    }
  }

  // Prerequisite -> Boolean
  // Checks if a prerequisite is a class or not
  isClass(prereq) {
    if (!prereq) {
      debugger
    }
    Object.prototype.hasOwnProperty.call(prereq, 'subject');
  }

  // Append mainClass to the prereqsFor array of prereqClass
  parseClass(mainClass, prereqClass, type) {
    const find = Keys.create({
      host: mainClass.host,
      termId: mainClass.termId,
      subject: prereqClass.subject,
      classUid: prereqClass.classUid,
    }).getHash();

    let found = {};
    for (const aClass of this.termDump.classes) {
      if (Keys.create(aClass).getHash() === find) {
        found = aClass;
        break;
      }
    }

    if (type === 'and') {
      if (found.prereqsFor === undefined) {
        found.prereqsFor = [];
      }
      found.prereqsFor.push(mainClass);
    } else {
      if (found.optPrereqsFor === undefined) {
        found.optPrereqsFor = [];
      }
      found.optPrereqsFor.push(mainClass);
    }
  }
}

export default new AddPreRequisiteFor();
