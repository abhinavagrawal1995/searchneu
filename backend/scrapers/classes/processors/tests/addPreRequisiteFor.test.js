/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import addPreRequisiteFor from '../addPreRequisiteFor';
// import Keys from '../../../../../common/Keys';

describe('addPreRequisiteFor tests', () => {
  const cs2500 = {
    classId: '2500',
    termId: '201829',
    host: 'neu.edu',
  };

  const cs2510 = {
    prereqs: {
      type: 'or',
      values: [{
        subject: 'CS',
        classUid: '2500_699845913',
        classId: '2500',
      }],
    },
    coreqs: {
      type: 'or',
      values: [{
        subject: 'CS',
        classUid: '2511_803118792',
        classId: '2511',
      }],
    },
    classId: '2510',
    termId: '201830',
    subject: 'CS',
  };

  const termDump = {
    classes: [cs2500, cs2510],
    sections: [],
  };

  it('should load in termDump', () => {
    addPreRequisiteFor.termDump = termDump;
    expect(addPreRequisiteFor.termDump).toBe(termDump);
  });

  describe('parseClass tests', () => {
    const outputPreReqClass = cs2500;
    outputPreReqClass.prereqsFor = [];
    outputPreReqClass.prereqsFor.push(cs2510);
  });
});
