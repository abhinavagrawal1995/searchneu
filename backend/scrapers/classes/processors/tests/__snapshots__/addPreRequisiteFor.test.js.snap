// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addPreRequisiteFor tests should sort some optPrereqsFor 1`] = `
Object {
  "values": Array [
    Object {
      "classId": "5",
      "classUid": "2500_699845913",
      "subject": "CS",
    },
    Object {
      "classId": "7",
      "classUid": "2500_699845913",
      "subject": "EECE",
    },
    Object {
      "classId": "11",
      "classUid": "2500_699845913",
      "subject": "EECE",
    },
    Object {
      "classId": "2",
      "classUid": "2500_699845913",
      "subject": "MATH",
    },
    Object {
      "classId": "3",
      "classUid": "2500_699845913",
      "subject": "MATH",
    },
  ],
}
`;

exports[`addPreRequisiteFor tests should sort some prereqsFor 1`] = `
Object {
  "values": Array [
    Object {
      "classId": "7",
      "classUid": "2500_699845913",
      "subject": "EECE",
    },
    Object {
      "classId": "11",
      "classUid": "2500_699845913",
      "subject": "EECE",
    },
    Object {
      "classId": "5",
      "classUid": "2500_699845913",
      "subject": "CS",
    },
    Object {
      "classId": "2",
      "classUid": "2500_699845913",
      "subject": "MATH",
    },
    Object {
      "classId": "3",
      "classUid": "2500_699845913",
      "subject": "MATH",
    },
  ],
}
`;
